File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    static bool compare(pair<int, int> p1, pair<int, int> p2)
    {
        if (p1.second == p2.second)
            return p1.first > p2.first;
        return p1.second > p2.second;
    };
    vector<int> topKFrequent(vector<int>& nums, int k) 
    {
        vector<int> out;
        unordered_map<int, int> m;
        for(int num : nums)
        {
            m[num]++;
        }

        vector<pair<int, int>> freq(m.begin(), m.end());
        
        sort(freq.begin(), freq.end(), compare);

        for(int i = 0; i < k; i++)
        {
            out.push_back(freq[i].first);
        }
        return out;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cs

public class Solution {
    public int[] TopKFrequent(int[] nums, int k)
    {
        var dic = new Dictionary<int, int>();

        foreach (int num in nums)
        {
            if (dic.TryGetValue(num, out int val))
                dic[num]++;
            else
                dic[num] = 1;
        }

        var freq = new List<List<int>>();
        for (int i = 0; i < nums.Length + 1; i++)
            freq.Add(new List<int>());
        foreach (var pair in dic)
            freq[pair.Value].Add(pair.Key);

        List<int> result = new List<int>();
        for (int i = freq.Count - 1; k > 0; i--)
        {
            for(int j = 0; j < freq[i].Count & k > 0; j++)
            {
                result.Add(freq[i][j]);
                k--;
            }
        }

        return result.ToArray();
    }
}
